homeassistant/components/zwave_js/fan.py

```diff
@@ -225,3 +236,124 @@ def zwave_speed_to_percentage(self, zwave_speed: int) -> int:
         # the UI handles steps e.g., for a 3-speed fan, you get steps at 33,
         # 67, and 100.
         return round(percentage)
+
+
+class ZwaveThermostatFan(ZWaveBaseEntity, FanEntity):
+    """Representation of a Z-Wave thermostat fan."""
+
+    _fan_mode: ZwaveValue
+    _fan_off: ZwaveValue | None = None
+    _fan_state: ZwaveValue | None = None
+
+    def __init__(
+        self, config_entry: ConfigEntry, client: ZwaveClient, info: ZwaveDiscoveryInfo
+    ) -> None:
+        """Initialize the thermostat fan."""
+        super().__init__(config_entry, client, info)
+
+        self._fan_mode = self.get_zwave_value(
+            THERMOSTAT_FAN_MODE_PROPERTY,
+            CommandClass.THERMOSTAT_FAN_MODE,
+            add_to_watched_value_ids=True,
+            check_all_endpoints=True,
+        )
+
+        if not self._fan_mode:
+            raise ValueError("Thermostat fan mode is required")
+
+        self._fan_off = self.get_zwave_value(
+            THERMOSTAT_FAN_OFF_PROPERTY,
+            CommandClass.THERMOSTAT_FAN_MODE,
+            add_to_watched_value_ids=True,
+            check_all_endpoints=True,
+        )
+        self._fan_state = self.get_zwave_value(
+            THERMOSTAT_FAN_STATE_PROPERTY,
+            CommandClass.THERMOSTAT_FAN_STATE,
+            add_to_watched_value_ids=True,
+            check_all_endpoints=True,
+        )
+
+    async def async_turn_on(
+        self,
+        speed: str | None = None,
+        percentage: int | None = None,
+        preset_mode: str | None = None,
+        **kwargs: Any,
+    ) -> None:
+        """Turn the device on."""
+        if self._fan_off:
```

Comment:
Should there be something more than silence if this property does not exist? Does Z-Wave JS even allow this to happen (the property is from the same command class as the primary value)? If there's no ability for control, should a fan entity even exist?

The discovery schema does have a required values field to enforce that more than one value ID exists in order to be discovered.

Reply:
The `off` attribute was added at ccVersion 2, so devices that comply with earlier versions will not have it. I think it's still useful to have a fan entity to control the "fan mode". For example on my HVAC the `Auto low` mode is the same as "off" in practice (I haven't seen it turn on when it is in that mode). I don't think we can generalize this to all HVACs though which is why nothing happens when the `off` attribute isn't available.

Reply:
OK, thanks for pointing that out. (I think the "off" value is actually only [available](https://github.com/zwave-js/node-zwave-js/blob/b67a28be269de49f56f21b949ab677d2f22e8e68/packages/zwave-js/src/lib/commandclass/ThermostatFanModeCC.ts#L379-L384) if the Mode CC is >= 3, but that doesn't matter much).

So I have a thermostat that implements V1 CC. The toggle switch is always on, and if I toggle it just reverts from off to on. If I don't know the distinction between V1 V3, I would be confused.

@MartinHjelmare Should there be something that indicates to the user on/off is not possible? At least a log warning?

Reply:
Weird I was looking at this line for the version: https://github.com/zwave-js/node-zwave-js/blob/b67a28be269de49f56f21b949ab677d2f22e8e68/packages/zwave-js/src/lib/commandclass/ThermostatFanModeCC.ts#L333-L334

But all other checks are using >= 3, one of these is incorrect

Reply:
You can set off mode in V2, but the report does not include the off bit, that was added in V3. So the Value API only supports V3, but you can still use the CC API and set off mode.

Reply:
Please invert the check and raise HomeAssistantError with a message as exception argument. Then we can outdent below.